File: start-db-core/src/test/scala/org/urbcomp/start/db/executor/SparkExecutor.scala
/*
 * Copyright 2022 ST-Lab
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU General Public License version 2 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */

package org.urbcomp.start.db.executor

import org.apache.calcite.sql.{SqlDialect, SqlNode}
import org.apache.spark.sql.SparkSession
import org.urbcomp.start.db.util.{MetadataUtil, SqlParam}

class SparkExecutor {

  def execute(sql: SqlNode): Unit = {
    // 获取缓存的配置信息
    val param = SqlParam.CACHE.get()

    // TODO 获取数据

    val geomesaParam = Map("" -> "")
    val sparkSession: SparkSession = SparkSession
      .builder()
      .appName("start-db sql app")
      .master("local[*]")
      .getOrCreate()

    // 读取相关的表数据
    sparkSession.read
      .format("geomesa")
      .options(geomesaParam)
      .option("geomesa.feature", "chicago")
      .load()
      .createTempView("")

    // 基于sql对读取数据进行查询
    sparkSession.sql(sql.toSqlString(new SqlDialect(SqlDialect.EMPTY_CONTEXT)).getSql)

    // 写入到中间存储介质中
  }

}

File: cupid-db-core/src/main/java/org/urbcomp/cupid/db/spark/SparkExecutor.java
/* 
 * Copyright (C) 2022  ST-Lab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.urbcomp.cupid.db.spark;

import org.urbcomp.cupid.db.infra.MetadataResult;
import org.urbcomp.cupid.db.util.SparkSqlParam;

import java.util.concurrent.TimeoutException;

/**
 * @author jimo
 **/
public class SparkExecutor {

    /**
     * 提交到spark并返回结果
     * 如果是异步，那么只返回一个提交成功
     */
    public <T> MetadataResult<T> execute(SparkSqlParam param) {
        final ISparkSubmitter submitter = ISparkSubmitter.getInstance();
        final SubmitResult res = submitter.submit(param);
        if (param.isAsync()) {
            // TODO 记录执行任务到缓存队列
            return MetadataResult.buildDDLResult(0);
        }
        // wait result
        try {
            submitter.waitToFinish(res);
        } catch (TimeoutException e) {
            throw new RuntimeException("sync wait timeout", e);
        }
        return ISparkDataRead.getReader(param.getExportType()).read(res.getSqlId());
    }
}
